File: tracks/go/exercises/io-and-files/debugging/buggy.go
// Package reports provides utilities for writing and reading daily reports.
// There are 4 bugs in this file. Find them.
package reports

import (
	"bufio"
	"fmt"
	"io"
	"os"
	"path/filepath"
	"time"
)

// ============================================================================
// BUG 1 LIVES HERE
// writeReport writes a structured report to path.
// The file should contain the report content when the function returns.
// ============================================================================

// Report holds data for a daily operational report.
type Report struct {
	Date     time.Time
	Service  string
	Requests int
	Errors   int
	P99Ms    float64
}

// writeReport serializes r and writes it to the file at path.
// Creates or truncates the file. Returns an error if writing fails.
func writeReport(path string, r Report) error {
	f, err := os.Create(path)
	if err != nil {
		return fmt.Errorf("create %s: %w", path, err)
	}

	bw := bufio.NewWriter(f)
	// BUG 1: defer f.Close() is registered AFTER defer bw.Flush().
	// Defers run LIFO — so f.Close() runs FIRST, then bw.Flush() tries to
	// write to an already-closed file. The data sits in the buffer and never
	// reaches disk. The file is empty and no error is surfaced.
	defer bw.Flush() // registered first — runs LAST (after Close!)
	defer f.Close()  // registered second — runs FIRST (closes file before Flush)

	fmt.Fprintf(bw, "date:     %s\n", r.Date.Format("2006-01-02"))
	fmt.Fprintf(bw, "service:  %s\n", r.Service)
	fmt.Fprintf(bw, "requests: %d\n", r.Requests)
	fmt.Fprintf(bw, "errors:   %d\n", r.Errors)
	fmt.Fprintf(bw, "p99_ms:   %.2f\n", r.P99Ms)

	return nil
}

// ============================================================================
// BUG 2 LIVES HERE
// processLogLines reads lines from r, counts them, and returns any error.
// It should surface I/O errors that occur while reading.
// ============================================================================

// processLogLines counts lines in r, returning count and any error.
func processLogLines(r io.Reader) (int, error) {
	count := 0
	scanner := bufio.NewScanner(r)
	for scanner.Scan() {
		count++
		_ = scanner.Text()
	}
	// BUG 2: something is missing here.
	// When the reader errors mid-stream, this function returns nil instead of the error.
	return count, nil
}

// ============================================================================
// BUG 3 LIVES HERE
// buildReportPath constructs the file path for a report.
// Should work correctly on all operating systems, including Windows.
// ============================================================================

// buildReportPath returns the file path for a report given the base directory,
// service name, and date. The path is: base/service/YYYY-MM-DD.report
func buildReportPath(base, service string, date time.Time) string {
	// BUG 3: this concatenation uses a hardcoded separator.
	// On Unix it works. On Windows it produces paths the OS can't open.
	return base + "/" + service + "/" + date.Format("2006-01-02") + ".report"
}

// ============================================================================
// BUG 4 LIVES HERE
// processAll processes a list of report files, reading each in turn.
// Each file should be closed after it is processed, not deferred to the end.
// ============================================================================

// ProcessResult holds the outcome of processing one file.
type ProcessResult struct {
	Path  string
	Lines int
	Err   error
}

// processAll opens and reads each file in paths, returning results.
// Files must be closed after each is processed (not deferred to function return).
func processAll(paths []string) []ProcessResult {
	results := make([]ProcessResult, 0, len(paths))

	for _, path := range paths {
		f, err := os.Open(path)
		if err != nil {
			results = append(results, ProcessResult{Path: path, Err: err})
			continue
		}
		// BUG 4: resource leak. This defer runs when processAll returns,
		// not at the end of each loop iteration. With 1000 files, you'd
		// hold 1000 open file descriptors.
		defer f.Close()

		n, err := processLogLines(f)
		results = append(results, ProcessResult{Path: path, Lines: n, Err: err})
	}

	return results
}

// ============================================================================
// Helper: not buggy, used by tests
// ============================================================================

// readLines reads all lines from path, returning them as a slice of strings.
func readLines(path string) ([]string, error) {
	f, err := os.Open(path)
	if err != nil {
		return nil, fmt.Errorf("open: %w", err)
	}
	defer f.Close()

	var lines []string
	scanner := bufio.NewScanner(f)
	for scanner.Scan() {
		lines = append(lines, scanner.Text())
	}
	return lines, scanner.Err()
}

// validPath makes buildReportPath result cross-platform comparable in tests.
func validPath(base, service string, date time.Time) string {
	return filepath.Join(base, service, date.Format("2006-01-02")+".report")
}

File: tracks/rust/exercises/io-and-files/debugging/buggy.rs
// Log Processor — Debugging Exercise (Rust / I/O and Files)
//
// This code compiles (mostly) but has 4 bugs — one per function.
// Read the symptoms in debugging/README.md, find the root cause, and fix it.
//
// Run: rustc --test buggy.rs && ./buggy

use std::fs::{self, File, OpenOptions};
use std::io::{self, BufWriter, Write};
use std::path::{Path, PathBuf};

// ---- Bug 1: BufWriter silently discards data ----
//
// write_summary writes a report to a file. The function returns Ok(()),
// the file is created — but it's always 0 bytes. No error is ever returned.
//
// Hint: BufWriter accumulates writes in an 8 KB buffer. When does that buffer
// get written to the underlying file? What happens when BufWriter is dropped?

pub fn write_summary(path: &Path, lines: &[String]) -> io::Result<()> {
    let file = File::create(path)?;
    let mut writer = BufWriter::new(file);

    for line in lines {
        writeln!(writer, "{}", line)?;
    }

    Ok(())
    // BUG: the BufWriter is dropped here without flushing.
    // Drop calls flush() but any flush error is silently discarded.
    // Small writes that fit in the 8 KB buffer are never written to disk.
}

// ---- Bug 2: Reading a file byte-by-byte without BufReader ----
//
// scan_log_slow counts lines containing a search term. It works correctly
// but is catastrophically slow on large files.
//
// Hint: File implements Read. Read::lines() requires BufRead, not Read.
// What trait provides lines()? What's missing here?

pub fn scan_log_slow(path: &Path, term: &str) -> io::Result<usize> {
    use std::io::Read;

    let mut file = File::open(path)?;
    let mut content = String::new();
    // BUG: reading the entire file into memory at once defeats the purpose of streaming.
    // For a 10 GB log file this allocates 10 GB of RAM.
    // The fix: wrap file in BufReader and iterate with .lines() instead of read_to_string.
    file.read_to_string(&mut content)?;

    let count = content
        .lines()
        .filter(|line| line.contains(term))
        .count();

    Ok(count)
}

// ---- Bug 3: unwrap() in library code ----
//
// open_report_file opens or creates a report file. But it panics when the
// path doesn't exist or when permissions are denied — instead of returning
// an error that the caller can handle.
//
// Hint: Library functions must never panic on expected failure cases.
// What should replace unwrap() here?

pub fn open_report_file(path: &Path) -> File {
    // BUG: unwrap() here panics the entire process if the file can't be opened.
    // Library code must propagate errors — the caller decides how to handle them.
    OpenOptions::new()
        .create(true)
        .append(true)
        .open(path)
        .unwrap()
}

// ---- Bug 4: String concatenation for paths ----
//
// build_archive_path constructs a path to an archive file.
// On Windows, the separator is `\` but this code always uses `/`.
// On any OS, this approach is fragile — a trailing slash in base_dir
// would produce double slashes like `logs//2026-02-18.tar.gz`.
//
// Hint: What type in std::path is designed to compose paths safely?

pub fn build_archive_path(base_dir: &str, date: &str) -> String {
    // BUG: string concatenation for paths is platform-specific and fragile.
    // This breaks on Windows (wrong separator) and is sensitive to trailing slashes.
    base_dir.to_string() + "/" + date + ".tar.gz"
}

// ---- Tests ----
//
// Note on Bugs 3 and 4:
// These bugs change return types (File -> io::Result<File>, String -> PathBuf).
// The tests below are written for the FIXED signatures and will not compile until
// you apply the fix. That compile error is intentional — it shows you exactly what
// the type should be. Bugs 1 and 2 compile but produce wrong output at runtime.

#[cfg(test)]
mod tests {
    use super::*;

    // Test 1: write_summary must actually write content
    // RUNTIME BUG — compiles, but file is empty after the call
    #[test]
    fn test_write_summary_creates_content() {
        let dir = std::env::temp_dir().join("foundry_debug_test");
        fs::create_dir_all(&dir).unwrap();
        let path = dir.join("summary.txt");

        let lines = vec![
            "errors: 42".to_string(),
            "warnings: 7".to_string(),
            "processed: 1000".to_string(),
        ];

        write_summary(&path, &lines).unwrap();

        let content = fs::read_to_string(&path).unwrap();
        assert!(
            !content.is_empty(),
            "file is empty — BufWriter was not flushed before drop"
        );
        assert!(content.contains("errors: 42"));
        assert!(content.contains("processed: 1000"));

        fs::remove_file(&path).ok();
        fs::remove_dir(&dir).ok();
    }

    // Test 2: scan_log_slow — correct results, but catastrophically slow on large files
    // RUNTIME BUG — compiles, returns correct results, but reads entire file into RAM
    #[test]
    fn test_scan_log_slow_counts_correctly() {
        let dir = std::env::temp_dir().join("foundry_debug_scan");
        fs::create_dir_all(&dir).unwrap();
        let path = dir.join("app.log");

        fs::write(
            &path,
            "INFO  start\nERROR disk full\nINFO  retry\nERROR max retries\nINFO  done\n",
        )
        .unwrap();

        let count = scan_log_slow(&path, "ERROR").unwrap();
        assert_eq!(count, 2, "expected 2 ERROR lines");

        fs::remove_file(&path).ok();
        fs::remove_dir(&dir).ok();
    }

    // Test 3: open_report_file must return io::Result<File>, not panic
    // TYPE BUG — the test below requires open_report_file to return io::Result<File>.
    // It will not compile until you change the return type and remove unwrap().
    #[test]
    fn test_open_report_file_returns_result() {
        let dir = std::env::temp_dir().join("foundry_debug_report");
        fs::create_dir_all(&dir).unwrap();
        let path = dir.join("report.log");

        // After the fix: open_report_file returns io::Result<File>
        let result: io::Result<File> = open_report_file(&path);
        assert!(result.is_ok(), "expected Ok(File)");

        // A bad path must return Err, not panic
        let bad_path = Path::new("/nonexistent/deeply/nested/report.log");
        let bad_result: io::Result<File> = open_report_file(bad_path);
        assert!(bad_result.is_err(), "expected Err for missing parent directory");

        fs::remove_dir_all(&dir).ok();
    }

    // Test 4: build_archive_path must return PathBuf and use OS-native separators
    // TYPE BUG — the test below requires build_archive_path to return PathBuf.
    // It will not compile until you change the return type to PathBuf.
    #[test]
    fn test_build_archive_path_correct_separator() {
        // After the fix: build_archive_path returns PathBuf
        let path: PathBuf = build_archive_path("/var/log/archives", "2026-02-18");

        assert_eq!(
            path.file_name().and_then(|n| n.to_str()),
            Some("2026-02-18.tar.gz"),
            "file name component should be '2026-02-18.tar.gz'"
        );
        assert_eq!(
            path.parent().and_then(|p| p.to_str()),
            Some("/var/log/archives"),
            "parent directory should be '/var/log/archives'"
        );
    }

    #[test]
    fn test_build_archive_path_no_double_slash() {
        let path: PathBuf = build_archive_path("/var/log/archives/", "2026-02-18");
        let path_str = path.to_string_lossy();
        assert!(
            !path_str.contains("//"),
            "path must not contain double slash: {}",
            path_str
        );
    }
}

fn main() {
    println!("Fix the bugs, then run: rustc --test buggy.rs && ./buggy");
}
